### PR TITLE
refactor CI structure and support daily build

### PR DESCRIPTION
**Pull request by SamYuan1990**
in this PR.
1. we have many pipeline files specific for golang and testing....
hence, I want to change it to below settings as main entry for as CI trigger.
```
|- daily.yml for daily build
|- pr.yml
|- push.yml
|- release.yml
```
1.a IMO, in the feature, we should considering daily build for both base image and image for latest, but which will be in another PR.

2. support daily build for container.
3. btw, why should we have so many pipeline with single command as ... `go vet`?
4. nit fix on checkout action and OS to running CI.

the pr level CI will upgrade to https://github.com/sustainable-computing-io/kepler/actions/runs/6148194995?pr=924
the push level CI will upgrade to https://github.com/SamYuan1990/kepler/actions/runs/6148434127
daily build level build logs: https://github.com/SamYuan1990/kepler/actions/runs/6157468840

### CODE REVIEW

**Comment by SamYuan1990** [comment]
k.... invalid value workflow reference: workflows must be defined at the top level of the .github/workflows/ directory

**Comment by SamYuan1990** [comment]
waiting for daily build result and meanwhile for code review.

File: .github/workflows/developer_local.yml
Comment (sthaha) [inline .github/workflows/developer_local.yml:1]: nit(feel free to ignore): 
```suggestion
name: local dev env validation
```

File: .github/workflows/golang.yml
Comment (sthaha) [inline .github/workflows/golang.yml:19]: This can make use of the go-version from go.mod file see:
* https://github.com/actions/setup-go#getting-go-version-from-the-gomod-file 
* https://github.com/sustainable-computing-io/kepler-operator/blob/v1alpha1/.github/workflows/pr-checks.yaml#L13
*

**Comment by sthaha** [comment]
Very nice @SamYuan1990 . Thanks a lot!

> btw, why should we have so many pipeline with single command as ... go vet?

Ideally we should point all CI issues on  PR. So as long as the result points out all CI issue and not only by one, I think it fine to have all checks in a single job. in fact it would be great if we had a `make lint` which devs can run locally that validates all checks locally. 
`lint` target can call other PHONY targets such as vet, gofmt, etc ..

**Comment by SamYuan1990** [comment]
> Very nice @SamYuan1990 . Thanks a lot!
> 
> > btw, why should we have so many pipeline with single command as ... go vet?
> 
> Ideally we should point all CI issues on PR. So as long as the result points out all CI issue and not only by one, I think it fine to have all checks in a single job. in fact it would be great if we had a `make lint` which devs can run locally that validates all checks locally. `lint` target can call other PHONY targets such as vet, gofmt, etc ..

@sthaha ,  we have `make golint`, but ... https://github.com/sustainable-computing-io/kepler/pull/859 and https://github.com/sustainable-computing-io/kepler/issues/724 we keep the duplication for now.

File: .github/workflows/integration_test.yml
Comment (jiere) [inline .github/workflows/integration_test.yml:21]: Since we have aligned the go version to go.mod, shall we bump the go version in go.mod to latest then?
i.e 1.18->1.21

File: .github/workflows/push.yml
Comment (jiere) [inline .github/workflows/push.yml:12]: Since we have pull_request.yml to check candidate merge PRs, and we also have daily.yaml to build image and check daily, do we still need push.yml here?

File: .github/workflows/image.yml
Comment (jiere) [inline .github/workflows/image.yml:5]: IIRC, only image.yml here is active workflow, image_base.yml and image_pr.yml are inactive ones.
Please confirm with @rootfs, since your current PR is a workflow cleanup and refactoring.

File: .github/workflows/daily.yml
Comment (jiere) [inline .github/workflows/daily.yml:5]: the timestamp here is a little bit strange, you mean 9:12 UTC daily?

File: .github/workflows/integration_test.yml
Comment (sthaha) [reply .github/workflows/integration_test.yml:21]: How about in a different PR so that the changes introduced by this PR is minimal and can be merged?

Comment (rootfs) [reply .github/workflows/integration_test.yml:21]: +1 have a min PR

**Comment by sthaha** [comment]
@SamYuan1990 

> @sthaha , we have make golint, but ... https://github.com/sustainable-computing-io/kepler/pull/859 and https://github.com/sustainable-computing-io/kepler/issues/724 we keep the duplication for now.

Since I saw we have `make check` is it sufficient to have just one job that just calls `make check` ?

**Comment by SamYuan1990** [comment]
> @SamYuan1990
> 
> > @sthaha , we have make golint, but ... #859 and #724 we keep the duplication for now.
> 
> Since I saw we have `make check` is it sufficient to have just one job that just calls `make check` ?

@marceloamaral , @sthaha, just IMO I suppose golangci-lint as which integrates lots of tools in one and run by single command as container is good enough for most for case. For my understanding, the reason for duplication we have today is just to provide detail informations for people who would like to use go command line tool to checking and validate the code as linting. Hence a `make check` as alias of `make golint` ... seems not the point.
IMO the point is:
as a sample of code blocks `go:=the_code`
before it been merged, every lint checking should be passed.
hence all in one as golangci-lint save a lots of works from dev ops point of view, if we treat pipeline as code, but it may stopped in the middle or other inconvenience. The `make check` as alias of `make golint`, no matter impls by golint container or go command line tools, are the same here, as in pipeline which running in a sequence, and breaks in the middle. (just for example: if go vet fails, then we know nothing about go lint result)
having go vet, go lint etc in different job as current impl, has the convenience to get the failure of lint(once it happen) and running in parallel model, but inconvenience for dev ops as pipeline as code for lots of duplicate.
...
I suppose we need another PR or issue to balance it.

File: .github/workflows/daily.yml
Comment (SamYuan1990) [reply .github/workflows/daily.yml:5]: When we run it daily, in fact any random time with in a day works, hence I use Sep 12th as the day I submit this PR. the Sep to hours and 12 to min.
I don't suppose it's harmful to run go sec and scorecard daily.
From security point of view, as new CVE been reported and announce randomly at any time. A daily check is better than a weekly with a higher frequency.

File: .github/workflows/integration_test.yml
Comment (SamYuan1990) [reply .github/workflows/integration_test.yml:21]: Resolve as this should be in another PR.

File: .github/workflows/push.yml
Comment (SamYuan1990) [reply .github/workflows/push.yml:12]: Daily build and PR level build are two different things, why not?

File: .github/workflows/image.yml
Comment (SamYuan1990) [reply .github/workflows/image.yml:5]: + @jichenjc here in loop.
- for image_pr.yml, which is used to build PR level image, as @rootfs request in the history. For our feature branch test usage, hence I suppose we need to keep it, in case of we have a big PR for example libebpf or others which need test locally. as far as I remember, we have this feature to build PR branch's image and upload to container reg, then cross validation.
- for image_base.yml, @jichenjc , @rootfs , as github agent has it's own linux header makes linux-header fails install in a ubi base image during image build.( there is no rpm package for github agent's linux header as they are ubuntu with deb package management system) Hence, any idea for how we build and daily build base image to reduce any CVE at OS level? or some yum package as linux-header?

File: .github/workflows/push.yml
Comment (jiere) [reply .github/workflows/push.yml:12]: I mean push_request.yml and push.yml.

Comment (SamYuan1990) [reply .github/workflows/push.yml:12]: pr, push, daily are different things.
For each PR, 
- pr means before the PR been merged. For example, we run lint before each PR and after the PR been merged the code passed the lint hence we don't need to run another turn for lint after the merge at push level or daily level.
- push means after the PR been merged. At push level, as we refresh test coverage, and build the image base on latest code.
- daily(or any schedule) is a daily build on main/default/specific branch. Take build image as an example, as we don't have PR merge per day, hence we have the image daily build to update dependences to latest as daily build. Same for go sec and score card.... SBOM as it relays on some place to store the result, so currently we make it pre releases.(store with release page on github) further if we find a place, we can make it into daily build as well.

**Review by rootfs — APPROVED** [review]

**Comment by rootfs** [comment]
Cleaning up and consolidating the workflows is a good start. Let's take this one first.

**Pull request merged**
